src/agent/graph.py

```
@@ -38,20 +42,115 @@ class _PickleSerde:
     """
 
     def dumps_typed(self, obj: Any) -> tuple[str, bytes]:
-        return ("pickle", pickle.dumps(obj))
+        try:
+            return ("pickle", pickle.dumps(obj))
+        except (TypeError, pickle.PicklingError):
+            return ("pickle", pickle.dumps(_make_picklable(obj)))
 
     def loads_typed(self, data: tuple[str, bytes]) -> Any:
-        return pickle.loads(data[1])
+        result = pickle.loads(data[1])
```

Ruff (0.15.20)

[error] `pickle` and modules that wrap it can be unsafe when used to deserialize untrusted data, possible security issue

(S301)